genie-web/src/main/java/com/netflix/genie/web/agent/apis/rpc/v4/endpoints/GRpcAgentFileStreamServiceImpl.java

```diff
@@ -142,333 +151,630 @@ public GRpcAgentFileStreamServiceImpl(
-        final boolean removed = this.jobIdControlStreamMap.remove(jobId, controlStreamObserver);
-        if (!removed) {
-            log.warn("Could not remove observer for job: {}, not found in map", jobId);
-        }
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void contribute(final Info.Builder builder) {
+        this.controlStreamsManager.contribute(builder);
+        this.transferManager.contribute(builder);
     }
 
-    private void handleFileTransferChunk(
-        final FileTransferStreamObserver fileTransferStreamObserver,
-        final String streamId,
-        final ByteString data
-    ) {
+    // Manages control streams, in theory one for each agent connected to this node
+    private static final class ControlStreamManager implements InfoContributor {
+        private final Map<String, ControlStreamObserver> controlStreamMap = Maps.newHashMap();
+        private final Cache<String, DirectoryManifest> manifestCache;
+        private final JobDirectoryManifestProtoConverter converter;
+        private final Counter fileTansferCounter;
+        private final MeterRegistry registry;
+
+        private ControlStreamManager(
+            final JobDirectoryManifestProtoConverter converter,
+            final AgentFileStreamProperties properties,
+            final MeterRegistry registry
+        ) {
+            this.converter = converter;
+            this.manifestCache = Caffeine.newBuilder()
+                .expireAfterWrite(properties.getManifestCacheExpiration())
+                .build();
+            this.fileTansferCounter = registry.counter(TRANSFER_COUNTER);
 
-        // Remove observer from the set of transfers waiting to start.
-        final boolean removed = this.pendingTransferObserversSet.remove(fileTransferStreamObserver);
-        if (removed) {
-            log.debug("Removed observer for file stream: {} from 'pending' set", streamId);
+            this.registry = registry;
         }
 
-        // Remove buffer from the set of transfers waiting to start and move it to the 'in progress' map.
-        final StreamBuffer streamBufferFromPending = this.pendingTransferBuffersMap.remove(streamId);
-        if (streamBufferFromPending != null) {
-            log.debug("Moving buffer for file stream: {} from 'pending' to 'in progress'", streamId);
-            this.inProgressTransferBuffersMap.put(streamId, streamBufferFromPending);
-        }
+        private synchronized void requestFile(
+            final String jobId,
+            final String fileTransferId,
+            final String relativePath,
+            final int startOffset,
+            final int endOffset
+        ) throws GenieNotFoundException {
+
+            final ControlStreamObserver controlStreamObserver = this.controlStreamMap.get(jobId);
+            if (controlStreamObserver == null) {
+                throw new GenieNotFoundException("No active stream control stream for job: " + jobId);
```

Comment:
On the server side I've been trying to use the newer exceptions. In this case there's [NotFoundException](https://github.com/Netflix/genie/blob/master/genie-web/src/main/java/com/netflix/genie/web/exceptions/checked/NotFoundException.java).

Reason being eventually we need to get rid of the old `genie-common` and those exceptions are outdated anyway as they tie the HTTP error code to the exception rather than delegating to outside mapper which has more context.

Reply:
I'll update